typing/errortrace.mli

```diff
@@ -102,6 +103,30 @@ val incompatible_fields : string -> type_expr -> type_expr -> (desc, _) elt
 
 val swap_trace : 'variety t -> 'variety t
 
+(* The traces (['variety t]) are the core error types.  However, we bundle them
+   up into three "top-level" error types, which are used elsewhere:
+   [unification_error], [equality_error], and [moregen_error].  In the case of
+   [equality_error], this has to bundle in extra information; in general, it
+   distinguishes the three types of errors and allows us to distinguish traces
+   that are being built (or processed) from those that are complete and have
+   become the final error. *)
+
+type unification_error = { trace : unification t } [@@unboxed]
+
+type equality_error =
+  { trace : comparison t;
+    subst : (type_expr * type_expr) list }
+
+type moregen_error = { trace : comparison t } [@@unboxed]
+
+(* Wraps up the two different kinds of [comparison] errors in one type *)
+type comparison_error =
+  | Equality_error of equality_error
+  | Moregen_error  of moregen_error
+
```

Comment:
With distinct complete error types, I would have hoped that `Errortrace.flatten` is no longer necessary, and that the type of `expand_trace` can be amended to `partial trace -> complete error` (For example by inlining it in a smart constructor for the various error types). Have you looked in this direction?

Reply:
This is an excellent idea, and looking into it (and the next suggestion) has already caught that we weren't calling `expand_trace` in enough places, producing even more detailed error messages!  Pushing it through will be a little more complex than a purely local change, but well worth it.

Reply:
I have made this change, and it's looking great.  I'm chasing down a couple places where I'm not sure what the correct expansion behavior is, and once I've found that this commit should be good to go.